src/main/scala/edu/ie3/powerFactory2psdm/model/powerfactory/GridModel.scala

```diff
@@ -0,0 +1,100 @@
+/*
+ * © 2021. TU Dortmund University,
+ * Institute of Energy Systems, Energy Efficiency and Energy Economics,
+ * Research group Distribution grid planning and operation
+ */
+
+package edu.ie3.powerFactory2psdm.model.powerfactory
+
+import com.typesafe.scalalogging.LazyLogging
+import edu.ie3.powerFactory2psdm.exception.pf.{
+  GridConfigurationException,
+  MissingParameterException
+}
+import edu.ie3.powerFactory2psdm.model.powerfactory.RawGridModel.{
+  LineTypes,
+  Lines,
+  Nodes,
+  Switches,
+  Trafos2w
+}
+
+final case class GridModel(
+    nodes: List[Node],
+    lineTypes: List[LineType],
+    lines: List[Line],
+    switches: List[Switch]
+)
+
+object GridModel extends LazyLogging {
+  def build(rawGrid: RawGridModel): GridModel = {
+
+    val models =
+      (rawGrid.nodes ++ rawGrid.lines ++ rawGrid.switches ++ rawGrid.trafos2w).flatten
+    val modelIds: Iterable[String] = models.map {
+      case node: Nodes =>
+        node.id.getOrElse(
+          throw MissingParameterException(s"Node $node has no defined id")
+        )
+      case line: Lines =>
+        line.id.getOrElse(
+          throw MissingParameterException(s"Line $line has no defined id")
+        )
+      case lineType: LineTypes =>
+        lineType.id.getOrElse(
+          throw MissingParameterException(
+            s"Line type $lineType has no defined id"
+          )
+        )
+      case switch: Switches =>
+        switch.id.getOrElse(
+          throw MissingParameterException(s"Switch $switch has no defined id")
+        )
+      case trafo2w: Trafos2w =>
+        trafo2w.id.getOrElse(
+          throw MissingParameterException(
+            s"Transformer $trafo2w has no defined id"
+          )
+        )
+    }
+    val duplicateIds =
+      modelIds.groupBy(identity).collect { case (x, List(_, _, _*)) => x }
```

Comment:
Maybe...?

```suggestion
    val duplicateIds =
      modelIds.toSet
```

Reply:
Yes I expected that 😂  
Well I thought about sets but how would you sensibly extract those duplicated ids afterwards? 

I could go

```
    if (modelIds.toSet.size < modelIds.size) {
      val duplicateIds =
        modelIds.groupBy(identity).collect { case (x, List(_, _, _*)) => x }
      throw GridConfigurationException(
        s"Can't build grid as there are grid elements with duplicated ids: $duplicateIds"
      )
    }
```
but I thought building the set would be similarly expensive as extracting the duplicates and in the case I have any duplicates i would have to get them someway anyways so I just did it right away. 

Is there a simpler way that I miss?

Reply:
Well there is still a way to go with a `Set` or `Seq.distinct`:

```
scala> val duplicates = List(1, 2, 3, 3)
val duplicates: List[Int] = List(1, 2, 3, 3)

scala> val uniques = duplicates.distinct
val uniques: List[Int] = List(1, 2, 3)

scala> val duplicatesOnly = duplicates.diff(duplicates.distinct)
val duplicatesOnly: List[Int] = List(3)
```